tests/ui/unconditional_recursion.rs

```diff
@@ -158,6 +158,15 @@ struct S5;
 impl_partial_eq!(S5);
 //~^ ERROR: function cannot return without recursing
 
+struct S6;
+
+impl std::string::ToString for S6 {
+    fn to_string(&self) -> String {
+        //~^ ERROR: function cannot return without recursing
+        self.to_string()
```

Comment:
Perhaps add another test where `self` is put into a local before calling its `to_string`?

And another one calling `(self as &Self).to_string()` just for extra fun? 😋

Reply:
> Perhaps add another test where `self` is put into a local before calling its `to_string`?

For that one, we agree that it's just for cases like:

```rust
let x = self;
x.to_string()
```

and nothing else? Because otherwise it becomes quite tricky to check if `self` was mutated... However if you know of a way to check it, I'm very interested. :)

> And another one calling `(self as &Self).to_string()` just for extra fun? 😋

Sure. ^^'

Reply:
> However if you know of a way to check it, I'm very interested. :)

There's the [clippy_utils::expr_or_init](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.expr_or_init.html) function, which should be helpful for that.

@rustbot author

Reply:
Done!